### PR TITLE
Add kubernetes 1.24 provider for kubevirtci

### PR DESCRIPTION
**Pull request by brianmcarey**
These changes are based on the automated PR: https://github.com/kubevirt/kubevirtci/pull/792

Updates on top of the above PR include:
- Bump CNAO to v0.76.0 for the 1.24 provider
- Update the kubeadm.conf to use `kubeadm.k8s.io/v1beta3`
- Add tolerations to cni for control-plane taint
- Update cri-o to version 1.24

### CODE REVIEW

File: cluster-provision/k8s/1.24/k8s_provision.sh [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+# Delete conf files created by crio / podman
+# so calico will create the interfaces by its own according the right configuration.
+# See https://github.com/cri-o/cri-o/issues/2411#issuecomment-540006558
+# It should happen before crio start, see https://github.com/cri-o/cri-o/issues/4276
+# About podman see https://github.com/kubernetes/kubernetes/issues/107687
+rm -f /etc/cni/net.d/*
+
+systemctl daemon-reload
+systemctl enable crio && systemctl start crio
+systemctl enable kubelet && systemctl start kubelet
```
Comment (dankenigsberg) [inline cluster-provision/k8s/1.24/k8s_provision.sh:43]:
```suggestion
systemctl enable --now crio kubelet
```

Comment (brianmcarey) [reply cluster-provision/k8s/1.24/k8s_provision.sh]: Good suggestion - updated now.

File: cluster-provision/k8s/1.24/manifests/cni.do-not-change.yaml [hunk 2]
```diff
@@ -3818,6 +3818,8 @@ spec:
           operator: Exists
         - key: node-role.kubernetes.io/master
           effect: NoSchedule
+        - key: node-role.kubernetes.io/control-plane
```
Comment (oshoval) [inline cluster-provision/k8s/1.24/manifests/cni.do-not-change.yaml:3821]: we should not change this file please, this is the original manifest
changes should be added via cni.diff

**Review by oshoval — COMMENTED** [review]
Thanks

**Comment by oshoval** [comment]
In order to ease the review, we usually make the first commit show the differences between the latest k8s folder,
can we do it here as well please ? for example by copying k8s-1.23 to k8s-1.24 and then apply the current first commit

**Comment by brianmcarey** [comment]
> In order to ease the review, we usually make the first commit show the differences between the latest k8s folder, can we do it here as well please ? for example by copying k8s-1.23 to k8s-1.24 and then apply the current first commit

I think the kubevirtci-bumper is doing just that - copying the previous provider and bumping the version. https://github.com/kubevirt/project-infra/blob/d7cf80b4872a241a7bffca05e1f018249c08ad6d/robots/pkg/kubevirtci/kubevirtci.go#L101

I have placed my changes on top in separate commits to try and help the review.

File: cluster-provision/k8s/1.24/manifests/cnao/network-addons-config-example.cr.yaml [hunk 3]
```diff
@@ -9,5 +9,4 @@ spec:
   linuxBridge: {}
   macvtap: {}
   multus: {}
-  nmstate: {}
```
[the remark this answers is not in the document]
Comment (brianmcarey) [reply cluster-provision/k8s/1.24/manifests/cnao/network-addons-config-example.cr.yaml]: I did not install standalone k8snm state - nmstate was removed as part of the CNAO bump commit. Will standalone k8snm state be required?

File: cluster-provision/k8s/1.24/manifests/cni.do-not-change.yaml (hunk 2 above)
Comment (brianmcarey) [reply cluster-provision/k8s/1.24/manifests/cni.do-not-change.yaml:3821]: yes definitely - I will try to apply this to the cni.diff file.

**Comment by oshoval** [comment]
> > In order to ease the review, we usually make the first commit show the differences between the latest k8s folder, can we do it here as well please ? for example by copying k8s-1.23 to k8s-1.24 and then apply the current first commit
> 
> I think the kubevirtci-bumper is doing just that - copying the previous provider and bumping the version. https://github.com/kubevirt/project-infra/blob/d7cf80b4872a241a7bffca05e1f018249c08ad6d/robots/pkg/kubevirtci/kubevirtci.go#L101
> 
> I have placed my changes on top in separate commits to try and help the review.

I understand, it is just weird that if you look on `cluster-provision/k8s/1.24/manifests/cni.diff` it shows lots of diffs, where it should basically have only a few if any compared to k8s-1.23
(for example it already has `assign_ipv4": "true"` and so on, so not sure they should appear

File: cluster-provision/k8s/1.24/manifests/cnao/operator.yaml [hunk 4]
```diff
@@ -130,35 +130,36 @@ spec:
   template:
     metadata:
       annotations:
-        description: cluster-network-addons-operator manages the lifecycle of different Kubernetes network components on top of Kubernetes cluster
+        description: cluster-network-addons-operator manages the lifecycle of different
+          Kubernetes network components on top of Kubernetes cluster
```
Comment (dhiller) [inline cluster-provision/k8s/1.24/manifests/cnao/operator.yaml:134]: I guess that this is result of bumping cnao, so you might not be responsible for this, but shouldn't this be literal style, i.e a `|` in the end of L133?

**Review by dhiller — APPROVED** [review]
/approve

/hold

one question/nit inline, feel free to unhold any time :)

File: cluster-provision/k8s/1.24/manifests/cni.diff [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
++            - name: CALICO_IPV6POOL_NAT_OUTGOING
++              value: "true"
+           securityContext:
+             privileged: true
+           resources:
+@@ -3818,6 +3818,8 @@ spec:
+           operator: Exists
+         - key: node-role.kubernetes.io/master
+           effect: NoSchedule
++        - key: node-role.kubernetes.io/control-plane
```
Comment (oshoval) [inline cluster-provision/k8s/1.24/manifests/cni.diff:82]: Thanks for the fix
btw, maybe latest calico has this already built in ? 
but we can bump on other PR if needed and try

File: cluster-provision/k8s/1.24/manifests/cni.do-not-change.yaml [hunk 6]
```diff
@@ -3818,8 +3818,6 @@ spec:
           operator: Exists
         - key: node-role.kubernetes.io/master
           effect: NoSchedule
-        - key: node-role.kubernetes.io/control-plane
```
Comment (oshoval) [inline cluster-provision/k8s/1.24/manifests/cni.do-not-change.yaml]: I think it should be squashed with the commit that presented it please

File: cluster-provision/k8s/1.24/k8s_provision.sh [hunk 7]
```diff
@@ -55,9 +55,11 @@ cat <<EOF >/etc/NetworkManager/conf.d/002-dhclient.conf
 dhcp=dhclient
 EOF
 
-sysctl -w net.netfilter.nf_conntrack_max=1000000
+
```
Comment (oshoval) [inline cluster-provision/k8s/1.24/k8s_provision.sh:58]: can you please drop this empty line ?

File: cluster-provision/k8s/1.24/manifests/cnao/operator.yaml (hunk 4 above)
Comment (brianmcarey) [reply cluster-provision/k8s/1.24/manifests/cnao/operator.yaml:134]: Yes its a result of bumping cnao[1] - I guess it should be fixed there as we are just consuming it here?

[1] https://github.com/kubevirt/cluster-network-addons-operator/blob/6130d3e299a658e75989c6f4cb81b805242b4664/manifests/cluster-network-addons/0.76.0/operator.yaml#L133

File: cluster-provision/k8s/1.24/k8s_provision.sh [hunk 8]
```diff
@@ -55,9 +55,11 @@ cat <<EOF >/etc/NetworkManager/conf.d/002-dhclient.conf
...
 echo "net.netfilter.nf_conntrack_max=1000000" >> /etc/sysctl.conf
 
+sysctl --system
```
Comment (oshoval) [inline cluster-provision/k8s/1.24/k8s_provision.sh:61]: maybe better to move this glued to line 59
since they belong

**Review by oshoval — COMMENTED** [review]
Thanks
last nits

File: cluster-provision/k8s/1.24/k8s_provision.sh (hunk 8 above)
Comment (brianmcarey) [reply cluster-provision/k8s/1.24/k8s_provision.sh]: Tidied up this and the empty line.

**Review by oshoval — APPROVED** [review]
Thanks

**Review by oshoval — APPROVED** [review]
Thanks
Great work

**Comment by oshoval** [comment]
There is still hold

btw maybe worth to bump on a follow PR CNAO on all other providers, so we will be aligned
there is a script for it `hack/bump-cnao.sh`

File: cluster-provision/k8s/1.24/manifests/kubeadm.conf [hunk 9]
```diff
@@ -28,20 +28,17 @@ apiServer:
...
 controllerManager:
   extraArgs:
-    feature-gates: IPv6DualStack=true
     node-cidr-mask-size-ipv6: "116"
 dns:
   type: CoreDNS
 etcd:
   local:
     dataDir: /var/lib/etcd
-featureGates:
```
Comment (xpivarc) [inline cluster-provision/k8s/1.24/manifests/kubeadm.conf]: Please include this change in the commit message. Something like "IPv6 Dual stack is default now".

File: cluster-provision/k8s/1.24/k8s_provision.sh [hunk 10]
```diff
@@ -55,9 +55,11 @@ cat <<EOF >/etc/NetworkManager/conf.d/002-dhclient.conf
 dhcp=dhclient
 EOF
 
-sysctl -w net.netfilter.nf_conntrack_max=1000000
```
Comment (xpivarc) [inline cluster-provision/k8s/1.24/k8s_provision.sh]: Can we leave a note about why do we make this change?

**Review by xpivarc — COMMENTED** [review]
Looks good, I have just 2 small requests.

**Comment by brianmcarey** [comment]
> There is still hold
> 
> btw maybe worth to bump on a follow PR CNAO on all other providers, so we will be aligned #805 what was the reason of the CNAO bump ?

The older CNAO version that is being used by the previous providers did not have tolerations for the taint node-role.kubernetes.io/control-plane:NoSchedule - the related pods were not getting scheduled.

File: cluster-provision/k8s/1.24/manifests/kubeadm.conf (hunk 9 above)
Comment (brianmcarey) [reply cluster-provision/k8s/1.24/manifests/kubeadm.conf]: Added note to the following commit message - https://github.com/kubevirt/kubevirtci/pull/800/commits/809e0476f9061a2dc510553436f25b197efc9606

File: cluster-provision/k8s/1.24/k8s_provision.sh (hunk 10 above)
Comment (brianmcarey) [reply cluster-provision/k8s/1.24/k8s_provision.sh]: Yes it happened again here on the cluster - https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/800/check-provision-k8s-1.24/1531922243037171712#1:build-log.txt%3A13068 - I tried looking into this but I couldn't find any reason for it.

Comment (oshoval) [reply cluster-provision/k8s/1.24/k8s_provision.sh]: I think that one of the packages we install creates it as well, so maybe the order changed in a subtle way causing it ?
maybe revert all the nice to have cosmetics, keep only the mandatory changes

**Comment by xpivarc** [comment]
/lgtm
@dhiller Do you want a final look?

File: cluster-provision/k8s/1.24/manifests/cnao/operator.yaml (hunk 4 above)
Comment (dhiller) [reply cluster-provision/k8s/1.24/manifests/cnao/operator.yaml:134]: Yes exactly.

File: cluster-provision/k8s/1.24/provision.sh [hunk 11]
```diff
@@ -99,6 +99,8 @@ enabled=1
 EOF
 dnf install -y cri-o
 
+systemctl enable --now crio
```
Comment (dhiller) [inline cluster-provision/k8s/1.24/provision.sh:102]: Why do we enable crio at two places, this and [that one](https://github.com/kubevirt/kubevirtci/pull/800/commits/809e0476f9061a2dc510553436f25b197efc9606#diff-e27be89aa116a14f88ff53f83b384465a6b1d8af5c9e5615b75e136266544b51R42)?

**Review by dhiller — COMMENTED** [review]
Looks good to me in general, one question inline.

File: cluster-provision/k8s/1.24/provision.sh (hunk 11 above)
Comment (oshoval) [reply cluster-provision/k8s/1.24/provision.sh:102]: Note the 2nd one is needed after deleting the CNI folder
https://github.com/kubevirt/kubevirtci/pull/800/commits/809e0476f9061a2dc510553436f25b197efc9606#diff-e27be89aa116a14f88ff53f83b384465a6b1d8af5c9e5615b75e136266544b51R37

(not sure about the first one indeed)

Comment (brianmcarey) [reply cluster-provision/k8s/1.24/provision.sh:102]: For the first one - cri-o needs to be running in order to pull the kubernetes images [here](https://github.com/kubevirt/kubevirtci/pull/800/files/f69b854f5981e792436640d5d2d25eb4db3c85c7..809e0476f9061a2dc510553436f25b197efc9606#diff-531d46398155ece227284b4b8c0bed55f400f96da4e41289f5dbec168cc607dbR142)

Comment (dhiller) [reply cluster-provision/k8s/1.24/provision.sh:102]: I just found it strange. I think that duplication in general is a code smell that points towards some common functionality. OTOH as this is a very small occurence which I can live with.

IIRC "someone" (citation needed tm) once said, avoid premature generalization, before you've had enough cases to see the commonalities clearly.

Thanks for the explanation!

Comment (dhiller) [reply cluster-provision/k8s/1.24/provision.sh:102]: To be more precise: there's a whole chapter about reuse in [Facts And Fallacies of Software Engineering](https://www.goodreads.com/book/show/83792.Facts_and_Fallacies_of_Software_Engineering) about it, very much worth reading :)

**Review by dhiller — APPROVED** [review]
/approve

Thank you! :tada:

**Pull request merged**